nucliadb_node/nucliadb_node/app.py

```diff
@@ -73,63 +60,47 @@ async def main() -> App:
     if node is None:
         raise Exception("No Key defined")
 
-    tracer_provider = get_telemetry(SERVICE_NAME)
-    if tracer_provider is not None:  # pragma: no cover
-        set_global_textmap(B3MultiFormat())
-        await init_telemetry(tracer_provider)  # To start asyncio task
-
-    logger.info(f"Node ID : {node}")
-
     worker = Worker(writer=writer, reader=reader, node=node)
     await worker.initialize()
 
-    grpc_finalizer = await start_grpc(writer=writer, reader=reader)
+    return worker
 
-    logger.info(f"======= Node sidecar started ======")
 
-    app = App()
-    app.worker = worker
-    app.writer = writer
-    app.reader = reader
+async def main():
+    writer = Writer(settings.writer_listen_address)
+    reader = Reader(settings.reader_listen_address)
+    worker = await start_worker(writer, reader)
 
-    # Using ensure_future as Signal handlers
-    # cannot handle couroutines as callbacks
-    def stop_pulling():
-        logger.info("Received signal to stop pulling!")
-        asyncio.ensure_future(worker.finalize())
+    tracer_provider = get_telemetry(SERVICE_NAME)
+    if tracer_provider is not None:  # pragma: no cover
+        set_global_textmap(B3MultiFormat())
+        await init_telemetry(tracer_provider)  # To start asyncio task
 
-    loop = asyncio.get_event_loop()
-    loop.add_signal_handler(signal.SIGUSR1, stop_pulling)
+    logger.info(f"Node ID : {worker.node}")
 
-    app.finalizers = [grpc_finalizer, worker.finalize]
+    grpc_finalizer = await start_grpc(writer=writer, reader=reader)
 
-    return app
+    logger.info(f"======= Node sidecar started ======")
 
+    metrics_server = await serve_metrics()
 
-def _cancel_all_tasks(loop):
-    to_cancel = tasks.all_tasks(loop)
-    if not to_cancel:
-        return
+    finalizers = [
+        grpc_finalizer,
+        worker.finalize,
+        metrics_server.shutdown,
+        writer.close,
+        reader.close,
+    ]
 
-    for task in to_cancel:
-        task.cancel()
+    if get_ff().enabled("nucliadb_node_sync_fs", default=False):
+        from nucliadb_node import write_sync
 
-    loop.run_until_complete(tasks.gather(*to_cancel, return_exceptions=True))
+        finalizers.append(write_sync.start())
```

Comment:
Is a write_sync.start() a finalizer on purpose to sync any left unsynced changes?

Reply:
it could be eventually but for us and right now, it's only just to tell the coroutine to exit.